File: python/1_Two_Sum.py
'''
Key:
1. using a hashmap to store numbers, and compare complements with numbers in hashmap
2. Reversing index and value, because of map.get() 
'''

from typing import List

class Solution:
    # input: int_list, int; output: int_list
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # range(start, stop, step)
        for i in range(len(nums)):  
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]
        return []


output = Solution().twoSum(nums = [3,2,4], target = 6)
print(output)

File: javascript/1_Two_Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
// 1.Loop
var twoSum = function(nums, target) {
    for (let i=0; i<nums.length; i++) {
        for (let j=i+1; j<nums.length; j++) {
            if (nums[i] + nums[j] == target) {
                return [i, j];
            }
        }
    }
};

// 2. Hashmap
var twoSum2 = function(nums, target) {
    const map = new Map();  // create a map (a collection of key-value pairs)
    for(let i=0; i<nums.length; i++) {
        const complement = target - nums[i];
        if(map.has(complement)) {
            return [map.get(complement), i];
        }
        map.set(nums[i], i);  // key: nums[i], value: i
    }
    return [];
};